File: ios/SIOPApp/Sources/IdentityEditor.swift
import SIOPKit
import SwiftUI

/// Creating an identity, or inspecting, renaming and deleting one.
///
/// Only the names can be edited. The subject is derived from the key, so it
/// is shown but never offered as a field: a different key would be a
/// different identity.
struct IdentityEditor: View {
    enum Mode: Identifiable {
        case create(clientID: String)
        case inspect(SIOPIdentity)

        var id: String {
            switch self {
            case let .create(clientID): return "create:\(clientID)"
            case let .inspect(identity): return identity.id
            }
        }
    }

    @EnvironmentObject private var session: AuthenticationSession
    @Environment(\.dismiss) private var dismiss

    let mode: Mode
    var onCreate: (SIOPIdentity) -> Void = { _ in }

    @State private var label = ""
    @State private var note = ""
    @State private var loaded = false
    @State private var confirmsDeletion = false

    private var inspected: SIOPIdentity? {
        if case let .inspect(identity) = mode { return identity }
        return nil
    }

    private var clientID: String {
        switch mode {
        case let .create(clientID): return clientID
        case let .inspect(identity): return identity.clientID
        }
    }

    private var trimmedLabel: String {
        label.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    var body: some View {
        let title: LocalizedStringKey = inspected == nil ? "Create identity" : "Identity details"
        let confirm: LocalizedStringKey = inspected == nil ? "Create" : "Save"
        return NavigationStack {
            Form {
                Section("Name") {
                    TextField("e.g. Personal, Testing", text: $label)
                        .accessibilityIdentifier("identity-label")
                }
                Section("Note") {
                    TextField("What it is for, for your own reference", text: $note, axis: .vertical)
                        .lineLimit(2...5)
                        .accessibilityIdentifier("identity-note")
                }
                if let identity = inspected {
                    details(of: identity)
                } else {
                    Section {
                        Text(verbatim: clientID)
                            .font(.footnote.monospaced())
                            .textSelection(.enabled)
                    } header: {
                        Text("Created for this RP")
                    } footer: {
                        Text("Creates an RSA key for this RP and derives the sub from its public key.")
                    }
                }
            }
            .navigationTitle(title)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button(confirm) { save() }
                        .disabled(trimmedLabel.isEmpty)
                        .accessibilityIdentifier("save-identity")
                }
            }
            .alert("Delete this identity?", isPresented: $confirmsDeletion) {
                Button("Delete identity and key", role: .destructive) {
                    if let identity = inspected {
                        session.delete(identity)
                        dismiss()
                    }
                }
                Button("Cancel", role: .cancel) {}
            } message: {
                Text(verbatim: deletionMessage)
            }
            .modifier(ReportsProblems())
            .onAppear {
                guard !loaded else { return }
                loaded = true
                if let identity = inspected {
                    label = identity.label
                    note = identity.note
                }
            }
        }
    }

    @ViewBuilder
    private func details(of identity: SIOPIdentity) -> some View {
        let jwk = session.publicKeys[identity.id]
        Section {
            Group {
                if let jwk {
                    Text(verbatim: jwk.thumbprint())
                } else {
                    Text("Key unreadable. This identity cannot answer.")
                }
            }
            .font(.footnote.monospaced())
            .textSelection(.enabled)
        } header: {
            Text("sub · derived from the public key / not editable")
        } footer: {
            Text("Only the name and note change. The public key and sub stay as they are.")
        }

        Section("RP it answers") {
            Text(verbatim: identity.clientID)
                .font(.footnote.monospaced())
                .textSelection(.enabled)
        }

        if let jwk {
            Section {
                DisclosureGroup("Public key sub_jwk and the JSON it is derived from") {
                    Text(verbatim: Self.explanation(of: jwk))
                        .font(.caption.monospaced())
                        .textSelection(.enabled)
                }
            }
        }

        Section {
            Button("Delete identity and key", role: .destructive) {
                confirmsDeletion = true
            }
            .accessibilityIdentifier("delete-identity")
        }
    }

    private var deletionMessage: String {
        guard let identity = inspected else { return "" }
        let subject = session.subject(of: identity) ?? String(localized: "(key unreadable)")
        let consequence = String(localized: "The signing key is deleted too. Unless the key is restored from a backup, this sub can never answer again. This does not delete the account the RP holds.")
        return "\(identity.displayName)\n\(subject)\n\n\(consequence)"
    }

    private func save() {
        let note = note.trimmingCharacters(in: .whitespacesAndNewlines)
        if let identity = inspected {
            session.relabel(identity, label: trimmedLabel, note: note)
            dismiss()
        } else if let created = session.createIdentity(for: clientID, label: trimmedLabel, note: note) {
            onCreate(created)
            dismiss()
        }
    }

    private static func explanation(of jwk: RSAPublicJWK) -> String {
        """
        sub_jwk
        {
          "kty": "\(jwk.kty)",
          "e": "\(jwk.e)",
          "n": "\(jwk.n)"
        }

        \(String(localized: "JSON the thumbprint is taken over"))
        \(jwk.canonicalJSON)

        SHA-256 → Base64url
        \(jwk.thumbprint())
        """
    }
}

File: android/app/src/main/kotlin/jp/co/pendako/siop/app/IdentityEditor.kt
package jp.co.pendako.siop.app

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import jp.co.pendako.siop.RsaPublicJwk
import jp.co.pendako.siop.SiopIdentity

sealed interface EditorMode {
    data class Create(val clientId: String) : EditorMode
    data class Inspect(val identity: SiopIdentity) : EditorMode
}

/**
 * Creating an identity, or inspecting, renaming and deleting one.
 *
 * Only the names can be edited. The subject is derived from the key, so it is
 * shown but never offered as a field: a different key would be a different
 * identity.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun IdentityEditor(
    session: AuthenticationSession,
    mode: EditorMode,
    onDismiss: () -> Unit,
    onCreate: (SiopIdentity) -> Unit = {},
) {
    val inspected = (mode as? EditorMode.Inspect)?.identity
    val clientId = when (mode) {
        is EditorMode.Create -> mode.clientId
        is EditorMode.Inspect -> mode.identity.clientId
    }
    var label by rememberSaveable { mutableStateOf(inspected?.label ?: "") }
    var note by rememberSaveable { mutableStateOf(inspected?.note ?: "") }
    var confirmsDeletion by remember { mutableStateOf(false) }
    val trimmedLabel = label.trim()

    fun save() {
        val trimmedNote = note.trim()
        if (inspected != null) {
            session.relabel(inspected, trimmedLabel, trimmedNote)
            onDismiss()
        } else {
            session.createIdentity(clientId, trimmedLabel, trimmedNote)?.let {
                onCreate(it)
                onDismiss()
            }
        }
    }

    Dialog(onDismissRequest = onDismiss, properties = DialogProperties(usePlatformDefaultWidth = false)) {
        Surface(Modifier.fillMaxSize()) {
            Scaffold(
                topBar = {
                    TopAppBar(
                        title = {
                            Text(stringResource(if (inspected == null) R.string.create_identity else R.string.identity_details))
                        },
                        navigationIcon = {
                            TextButton(onClick = onDismiss) { Text(stringResource(R.string.cancel)) }
                        },
                        actions = {
                            TextButton(
                                onClick = ::save,
                                enabled = trimmedLabel.isNotEmpty(),
                                modifier = Modifier.testTag("save-identity"),
                            ) {
                                Text(stringResource(if (inspected == null) R.string.create else R.string.save))
                            }
                        },
                    )
                },
            ) { padding ->
                Column(
                    modifier = Modifier
                        .padding(padding)
                        .fillMaxSize()
                        .verticalScroll(rememberScrollState())
                        .padding(16.dp),
                    verticalArrangement = Arrangement.spacedBy(20.dp),
                ) {
                    OutlinedTextField(
                        value = label,
                        onValueChange = { label = it },
                        label = { Text(stringResource(R.string.name)) },
                        placeholder = { Text(stringResource(R.string.name_placeholder)) },
                        singleLine = true,
                        modifier = Modifier.fillMaxWidth().testTag("identity-label"),
                    )
                    OutlinedTextField(
                        value = note,
                        onValueChange = { note = it },
                        label = { Text(stringResource(R.string.note)) },
                        placeholder = { Text(stringResource(R.string.note_placeholder)) },
                        minLines = 2,
                        maxLines = 5,
                        modifier = Modifier.fillMaxWidth().testTag("identity-note"),
                    )
                    if (inspected != null) {
                        Details(session, inspected, onDelete = { confirmsDeletion = true })
                    } else {
                        Section(
                            header = stringResource(R.string.created_for_rp),
                            footer = stringResource(R.string.creates_key_footer),
                        ) {
                            SelectionContainer {
                                Text(clientId, style = MaterialTheme.typography.bodySmall.mono())
                            }
                        }
                    }
                }
            }
        }
    }

    if (confirmsDeletion && inspected != null) {
        val subject = session.subjectOf(inspected) ?: stringResource(R.string.key_unreadable_paren)
        val consequence = stringResource(R.string.delete_consequence)
        AlertDialog(
            onDismissRequest = { confirmsDeletion = false },
            title = { Text(stringResource(R.string.delete_confirm_title)) },
            text = { Text("${inspected.displayName}\n$subject\n\n$consequence") },
            confirmButton = {
                TextButton(
                    onClick = {
                        confirmsDeletion = false
                        session.delete(inspected)
                        onDismiss()
                    },
                    colors = ButtonDefaults.textButtonColors(contentColor = Palette.danger),
                    modifier = Modifier.testTag("confirm-delete"),
                ) {
                    Text(stringResource(R.string.delete_identity_and_key))
                }
            },
            dismissButton = {
                TextButton(onClick = { confirmsDeletion = false }) { Text(stringResource(R.string.cancel)) }
            },
        )
    }
}

@Composable
private fun Details(session: AuthenticationSession, identity: SiopIdentity, onDelete: () -> Unit) {
    val jwk = session.publicKeys[identity.id]
    var showsKey by rememberSaveable { mutableStateOf(false) }

    Section(
        header = stringResource(R.string.sub_header),
        footer = stringResource(R.string.only_names_change),
    ) {
        SelectionContainer {
            Text(
                jwk?.thumbprint() ?: stringResource(R.string.key_unreadable_detail),
                style = MaterialTheme.typography.bodySmall.mono(),
            )
        }
    }

    Section(header = stringResource(R.string.rp_it_answers)) {
        SelectionContainer {
            Text(identity.clientId, style = MaterialTheme.typography.bodySmall.mono())
        }
    }

    if (jwk != null) {
        Column {
            TextButton(onClick = { showsKey = !showsKey }) {
                Text(stringResource(R.string.public_key_disclosure))
            }
            if (showsKey) {
                SelectionContainer {
                    Text(explanation(jwk, stringResource(R.string.thumbprint_json)), style = MaterialTheme.typography.labelSmall.mono())
                }
            }
        }
    }

    OutlinedButton(
        onClick = onDelete,
        colors = ButtonDefaults.outlinedButtonColors(contentColor = Palette.danger),
        modifier = Modifier.testTag("delete-identity"),
    ) {
        Text(stringResource(R.string.delete_identity_and_key))
    }
}

@Composable
private fun Section(header: String, footer: String? = null, content: @Composable () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Text(header, style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
        content()
        if (footer != null) {
            Text(footer, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }
    }
}

private fun explanation(jwk: RsaPublicJwk, canonicalCaption: String): String = """
    |sub_jwk
    |{
    |  "kty": "${jwk.kty}",
    |  "e": "${jwk.e}",
    |  "n": "${jwk.n}"
    |}
    |
    |$canonicalCaption
    |${jwk.canonicalJson}
    |
    |SHA-256 → Base64url
    |${jwk.thumbprint()}
""".trimMargin()
